File: zenscript-code-model/src/main/java/raylras/zen/model/type/NumberType.java
package raylras.zen.model.type;

import raylras.zen.model.CompilationEnvironment;
import raylras.zen.model.symbol.Operator;
import raylras.zen.model.symbol.Symbol;
import raylras.zen.model.symbol.SymbolFactory;
import raylras.zen.model.symbol.SymbolProvider;

import java.util.List;

public sealed interface NumberType extends Type, SymbolProvider
        permits ByteType, DoubleType, FloatType, IntType, LongType, ShortType {

    @Override
    default boolean isCastableTo(Type type, CompilationEnvironment env) {
        if (type instanceof NumberType || type instanceof StringType) {
            return true;
        }
        return Type.super.isCastableTo(type, env);
    }

    @Override
    default List<Symbol> getSymbols() {
        return SymbolFactory.builtinSymbols()
                .operator(Operator.NEG, this)
                .operator(Operator.ADD, this, params -> params.parameter("value", this))
                .operator(Operator.SUB, this, params -> params.parameter("value", this))
                .operator(Operator.MUL, this, params -> params.parameter("value", this))
                .operator(Operator.DIV, this, params -> params.parameter("value", this))
                .operator(Operator.MOD, this, params -> params.parameter("value", this))
                .operator(Operator.CONCAT, StringType.INSTANCE, params -> params.parameter("str", StringType.INSTANCE))
                .operator(Operator.EQUALS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.NOT_EQUALS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.LESS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.LESS_EQUALS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.GREATER, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.GREATER_EQUALS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .build();
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/type/NumberType.kt
package raylras.zen.model.type

import raylras.zen.model.CompilationEnvironment
import raylras.zen.model.symbol.Operator
import raylras.zen.model.symbol.Symbol
import raylras.zen.model.symbol.SymbolProvider
import raylras.zen.model.symbol.symbolSequence

interface NumberType : Type, SymbolProvider {
    override fun isCastableTo(that: Type?, env: CompilationEnvironment): Boolean {
        return when (that) {
            is NumberType, is StringType -> {
                true
            }

            else -> super.isCastableTo(that, env)
        }
    }

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        val builtin = symbolSequence {
            operator(Operator.NEG, this@NumberType)
            operator(Operator.ADD, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.SUB, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.MUL, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.DIV, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.MOD, this@NumberType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.CONCAT, StringType) {
                parameter("value", StringType)
            }
            operator(Operator.EQUALS, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.NOT_EQUALS, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.LESS, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.LESS_EQUALS, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.GREATER, BoolType) {
                parameter("value", this@NumberType)
            }
            operator(Operator.GREATER_EQUALS, BoolType) {
                parameter("value", this@NumberType)
            }
        }
        return builtin + getExpands(env)
    }
}
